server/src/main/java/io/crate/execution/ddl/tables/CreateTableRequest.java

```diff
@@ -83,21 +155,91 @@ public TimeValue ackTimeout() {
+
+            this.settings = readSettingsFromStream(in);
+            this.routingColumn = in.readOptionalString();
+            this.tableColumnPolicy = ColumnPolicy.VALUES.get(in.readVInt());
+            this.partitionedBy = in.readList(StreamInput::readStringList);
+            this.indices = in.readMap();
+
+            createIndexRequest = null;
             putIndexTemplateRequest = null;
         } else {
-            putIndexTemplateRequest = new PutIndexTemplateRequest(in);
-            createIndexRequest = null;
+            if (in.readBoolean()) {
+                createIndexRequest = new CreateIndexRequest(in);
+                putIndexTemplateRequest = null;
+            } else {
+                putIndexTemplateRequest = new PutIndexTemplateRequest(in);
+                createIndexRequest = null;
+            }
+            this.relationName = null;
+            this.colsToAdd = null;
+            this.pKeyIndices = null;
+            this.checkConstraints = null;
+            this.settings = null;
+            this.routingColumn = null;
+            this.tableColumnPolicy = null;
+            this.partitionedBy = null;
+            this.indices = null;
         }
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
-        boolean isIndexRequest = createIndexRequest != null;
-        out.writeBoolean(isIndexRequest);
-        MasterNodeRequest request = isIndexRequest ? createIndexRequest : putIndexTemplateRequest;
-        request.writeTo(out);
+        if (out.getVersion().onOrAfter(Version.V_5_4_0)) {
+            writeColsToAdd(out, relationName, checkConstraints, colsToAdd, pKeyIndices);
+            writeSettingsToStream(settings, out);
+            out.writeOptionalString(routingColumn);
+            out.writeVInt(tableColumnPolicy.ordinal());
+            out.writeCollection(partitionedBy, StreamOutput::writeStringCollection);
+            out.writeMap(indices);
+        } else {
+            boolean isIndexRequest = createIndexRequest != null;
+            out.writeBoolean(isIndexRequest);
+            MasterNodeRequest request = isIndexRequest ? createIndexRequest : putIndexTemplateRequest;
+            request.writeTo(out);
+        }
     }
+
+    public Settings settings() {
+        return settings;
+    }
+
+    @Nullable
+    public String routingColumn() {
+        return routingColumn;
+    }
+
+    public ColumnPolicy tableColumnPolicy() {
+        return tableColumnPolicy;
+    }
+
+    public List<List<String>> partitionedBy() {
+        return partitionedBy;
+    }
+
+    public Map<String, Object> indices() {
+        return indices;
+    }
+
+    public Map<String, String> checkConstraints() {
+        return this.checkConstraints;
+    }
+
+    public List<Reference> references() {
+        return this.colsToAdd;
+    }
+
+    public IntArrayList pKeyIndices() {
+        return this.pKeyIndices;
+    }
```

Comment:
I initialized all collections with nulls only because they are final and had to put some sentinel (anyway unused value) in BWC branch. Transport action doesn't treat them as nullable when handling 5.4 version requests.

But that's indeed confusing. I will update collections handling in transport to be nullable in 5.4 as well so that everything is aligned.

Reply:
RelationName and collections are @Nonnull in `AddColumnRequest`, which I'm going to extend. I will inti with empty collections then

Reply:
Made collections `Nonnull` instead since they are init as empty by default on analysis and mapping code already expects non-null for constraints.